File: Pandacap.LowLevel/Clients/ComputerVision.fs
namespace Pandacap.Clients

open System
open System.Threading
open Azure.AI.Vision.ImageAnalysis
open Azure.Identity

type ComputerVisionConfiguration = {
    Endpoint: string
    TenantId: string
}

type ComputerVisionProvider(config: ComputerVisionConfiguration) =
    member _.AnalyzeImageAsync(data: byte array, cancellationToken: CancellationToken) = task {
        let client = new ImageAnalysisClient(
            new Uri(config.Endpoint),
            new DefaultAzureCredential(
                new DefaultAzureCredentialOptions(
                    TenantId = config.TenantId)))

        let! analysis = client.AnalyzeAsync(
            BinaryData.FromBytes(data),
            VisualFeatures.Caption ||| VisualFeatures.Read,
            cancellationToken = cancellationToken)

        return String.concat "\n" [
            analysis.Value.Caption.Text

            for block in analysis.Value.Read.Blocks do
                for line in block.Lines do
                    line.Text
        ]
    }

File: Pandacap/ComputerVision.cs
using Azure.AI.Vision.ImageAnalysis;
using Azure.Identity;

namespace Pandacap
{
    public record ComputerVisionConfiguration(
        string Endpoint,
        string TenantId);

    public class ComputerVisionProvider(
        IEnumerable<ComputerVisionConfiguration> configs)
    {
        public async Task<string> AnalyzeImageAsync(byte[] data, CancellationToken cancellationToken)
        {
            if (configs.FirstOrDefault() is not ComputerVisionConfiguration config)
                return "";

            var client = new ImageAnalysisClient(
                new Uri(config.Endpoint),
                new DefaultAzureCredential(
                    new DefaultAzureCredentialOptions
                    {
                        TenantId = config.TenantId
                    }));

            var analysis = await client.AnalyzeAsync(
                BinaryData.FromBytes(data),
                VisualFeatures.Caption | VisualFeatures.Read,
                cancellationToken: cancellationToken);

            return string.Join("\n", [
                analysis.Value.Caption.Text,
                .. analysis.Value.Read.Blocks
                    .SelectMany(block => block.Lines)
                    .Select(block => block.Text)
            ]);
        }
    }
}
